src/Microsoft.AspNet.Mvc.Core/Routing/UrlHelper.cs

```diff
@@ -142,15 +138,59 @@ protected virtual string GeneratePathFromRoute(string routeName, RouteValueDicti
 
             // VirtualPathData.VirtualPath returns string.Empty for null.
             Debug.Assert(pathData.VirtualPath != null);
-
-            var fullPath = HttpContext.Request.PathBase.Add(pathData.VirtualPath).Value;
-            if (fullPath.Length == 0)
+            var pathBase = HttpContext.Request.PathBase;
+            if (!pathBase.HasValue)
             {
-                return "/";
+                if (pathData.VirtualPath.Length == 0)
+                {
+                    return "/";
+                }
+                else
+                {
+                    return !pathData.VirtualPath.StartsWith("/", StringComparison.Ordinal)
+                        ? "/" + pathData.VirtualPath
+                        : pathData.VirtualPath;
+                }
             }
             else
             {
-                return fullPath;
+                if (pathData.VirtualPath.Length == 0)
+                {
+                    return pathBase;
+                }
+                else
+                {
+                    var requiresLeadingSlash = !pathData.VirtualPath.StartsWith("/", StringComparison.Ordinal);
+                    var virtualPathStartIndex = 0;
+                    if (pathBase.Value.EndsWith("/", StringComparison.Ordinal))
+                    {
+                        if (!requiresLeadingSlash)
+                        {
+                            // If the base path has a trailing slash and the virtual path string has a leading slash,
+                            // we need to trim one of them.
+                            virtualPathStartIndex = 1;
+
+                        }
+
+                        requiresLeadingSlash = false;
+                    }
+
+                    var virtualPathLength = pathData.VirtualPath.Length - virtualPathStartIndex;
+
+                    var builder = new StringBuilder(
+                        pathBase.Value.Length
+                        + (requiresLeadingSlash ? 1 : 0)
+                        + virtualPathLength);
```

Comment:
If you're going to always use a stringbuilder here, you can just append pathbase, trim one character if needed, and then append the virtual path. The code would be much simpler.